File: src/polymarket/websocket_client.py
from __future__ import annotations

import asyncio
import inspect
import json
from typing import Any, Callable

import structlog
import websockets
from websockets.client import WebSocketClientProtocol

from src.config import Settings

logger = structlog.get_logger(__name__)


class PolymarketWebSocketClient:
    def __init__(self, settings: Settings):
        self.settings = settings
        self.ws_url = settings.polymarket_ws_url
        self.websocket: WebSocketClientProtocol | None = None
        self.message_handlers: dict[str, Callable] = {}
        self.running = False
        self._subscriptions: list[dict[str, Any]] = []

    def register_handler(self, message_type: str, handler: Callable):
        self.message_handlers[message_type] = handler

    async def connect(self):
        try:
            self.websocket = await websockets.connect(self.ws_url)
            logger.info("websocket_connected", url=self.ws_url)
            self.running = True
        except Exception as e:
            logger.error("websocket_connection_failed", error=str(e))
            raise

    async def subscribe_orderbook(self, market_id: str):
        if not self.websocket:
            await self.connect()

        message = {
            "type": "subscribe",
            "channel": "l2_book",
            "market": market_id,
        }
        await self.websocket.send(json.dumps(message))
        self._subscriptions.append(message)
        logger.info("orderbook_subscribed", market_id=market_id)

    async def subscribe_trades(self, market_id: str):
        if not self.websocket:
            await self.connect()

        message = {
            "type": "subscribe",
            "channel": "trades",
            "market": market_id,
        }
        await self.websocket.send(json.dumps(message))
        self._subscriptions.append(message)
        logger.info("trades_subscribed", market_id=market_id)

    async def _resubscribe_all(self):
        """Resubscribe to all channels after reconnection."""
        for message in self._subscriptions:
            try:
                await self.websocket.send(json.dumps(message))
                logger.info("resubscribed", channel=message.get("channel"), market=message.get("market"))
            except Exception as e:
                logger.error("resubscription_failed", error=str(e))

    async def listen(self):
        if not self.websocket:
            await self.connect()

        while self.running:
            try:
                message = await self.websocket.recv()
                data = json.loads(message)

                message_type = data.get("type")
                if message_type and message_type in self.message_handlers:
                    handler = self.message_handlers[message_type]
                    if inspect.iscoroutinefunction(handler):
                        await handler(data)
                    else:
                        handler(data)

            except websockets.exceptions.ConnectionClosed:
                logger.warning("websocket_connection_closed")
                await asyncio.sleep(5)
                await self.connect()
                await self._resubscribe_all()
            except Exception as e:
                logger.error("websocket_listen_error", error=str(e))
                await asyncio.sleep(1)

    async def close(self):
        self.running = False
        if self.websocket:
            await self.websocket.close()

File: src/polymarket/websocketClient.ts
import { on } from "node:events";
import WebSocket from "ws";
import type { Settings } from "../config.js";
import type { Logger } from "../logger.js";

type MessageHandler = (data: Record<string, unknown>) => void | Promise<void>;

export class PolymarketWebSocketClient {
  private ws: WebSocket | null = null;
  private readonly messageHandlers = new Map<string, MessageHandler>();
  running = false;

  constructor(
    private readonly settings: Settings,
    private readonly log: Logger,
  ) {}

  registerHandler(messageType: string, handler: MessageHandler): void {
    this.messageHandlers.set(messageType, handler);
  }

  async connect(): Promise<void> {
    try {
      this.ws = new WebSocket(this.settings.polymarketWsUrl);
      await new Promise<void>((resolve, reject) => {
        if (!this.ws) return reject(new Error("no ws"));
        this.ws.once("open", () => resolve());
        this.ws.once("error", reject);
      });
      this.log.info({ url: this.settings.polymarketWsUrl }, "websocket_connected");
      this.running = true;
    } catch (e) {
      this.log.error({ err: e }, "websocket_connection_failed");
      throw e;
    }
  }

  async subscribeOrderbook(marketId: string): Promise<void> {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) await this.connect();
    const message = { type: "subscribe", channel: "l2_book", market: marketId };
    this.ws!.send(JSON.stringify(message));
    this.log.info({ market_id: marketId }, "orderbook_subscribed");
  }

  async subscribeTrades(marketId: string): Promise<void> {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) await this.connect();
    const message = { type: "subscribe", channel: "trades", market: marketId };
    this.ws!.send(JSON.stringify(message));
    this.log.info({ market_id: marketId }, "trades_subscribed");
  }

  async listen(): Promise<void> {
    while (this.running) {
      if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
        try {
          await this.connect();
        } catch {
          await sleep(5000);
          continue;
        }
      }

      const ws = this.ws!;
      try {
        for await (const rawEvent of on(ws, "message") as AsyncIterableIterator<
          [WebSocket.RawData, boolean]
        >) {
          if (!this.running) return;

          const rawData = rawEvent[0];
          const raw = typeof rawData === "string" ? rawData : rawData.toString();

          let data: Record<string, unknown>;
          try {
            data = JSON.parse(raw) as Record<string, unknown>;
          } catch {
            continue;
          }

          const messageType = data.type;
          if (typeof messageType !== "string") continue;
          const handler = this.messageHandlers.get(messageType);
          if (handler) await handler(data);
        }
      } catch (e) {
        this.log.error({ err: e }, "websocket_listen_error");
      }

      if (!this.running) return;

      this.log.warn("websocket_connection_closed");
      await sleep(5000);
      try {
        await this.connect();
      } catch {
        /* next loop */
      }
    }
  }

  async close(): Promise<void> {
    this.running = false;
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((r) => setTimeout(r, ms));
}
